spec/unit/ssl/ssl_provider_spec.rb

```diff
@@ -0,0 +1,341 @@
+      int.sign(wrong_key, OpenSSL::Digest::SHA256.new)
+
+      expect {
+        subject.create_context(config.merge(cacerts: global_cacerts))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "Invalid signature for certificate '/CN=Test CA Subauthority'")
+    end
+
+    it 'raises if CRL signature for root CA is invalid', unless: Puppet::Util::Platform.jruby? do
+      crl = global_crls.first
+      crl.sign(wrong_key, OpenSSL::Digest::SHA256.new)
+
+      expect {
+        subject.create_context(config.merge(crls: global_crls))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "Invalid signature for CRL issued by '/CN=Test CA'")
+    end
+
+    it 'raises if CRL signature for intermediate CA is invalid', unless: Puppet::Util::Platform.jruby? do
+      crl = global_crls.last
+      crl.sign(wrong_key, OpenSSL::Digest::SHA256.new)
+
+      expect {
+        subject.create_context(config.merge(crls: global_crls))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "Invalid signature for CRL issued by '/CN=Test CA Subauthority'")
+    end
+
+    it 'raises if client cert is revoked' do
+      expect {
+        subject.create_context(config.merge(private_key: key('revoked-key.pem'), client_cert: cert('revoked.pem')))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "Certificate '/CN=revoked' is revoked")
+    end
+
+    it 'raises if intermediate issuer is missing' do
+      expect {
+        subject.create_context(config.merge(cacerts: [cert('ca.pem')]))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The issuer '/CN=Test CA Subauthority' of certificate '/CN=signed' cannot be found locally")
+    end
+
+    it 'raises if root issuer is missing' do
+      expect {
+        subject.create_context(config.merge(cacerts: [cert('intermediate.pem')]))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The issuer '/CN=Test CA' of certificate '/CN=Test CA Subauthority' is missing")
+    end
+
+    it 'raises if cert is not valid yet', unless: Puppet::Util::Platform.jruby? do
+      client_cert.not_before = Time.now + (5 * 60 * 60)
+      expect {
+        subject.create_context(config.merge(client_cert: client_cert))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The certificate '/CN=signed' is not yet valid, verify time is synchronized")
+    end
+
+    it 'raises if cert is expired', unless: Puppet::Util::Platform.jruby? do
+      client_cert.not_after = Time.at(0)
+      expect {
+        subject.create_context(config.merge(client_cert: client_cert))
+      }.to raise_error(Puppet::SSL::CertVerifyError,
+                       "The certificate '/CN=signed' has expired, verify time is synchronized")
+    end
+
+    it 'raises if crl is not valid yet', unless: Puppet::Util::Platform.jruby? do
+      future_crls = global_crls
```

Comment:
What is limiting these from being run on jruby?  Do we ever want/expect that to change?

Reply:
JRuby emulates ruby's openssl library (instead using the Java support), and it's a little buggy. This code is agent specific, so it's ok that this test doesn't pass on JRuby. I thought it was better to let all of the ssl specs run on jruby, and make exceptions where JRuby's openssl emulation is lacking.